src/main/java/pl/project13/maven/git/NativeGitProvider.java

```diff
@@ -515,4 +515,38 @@ private void runProcess(
       }
     }
   }
+
+  @Override
+  public AheadBehind getAheadBehind() throws GitCommitIdExecutionException {
+    try {
+      Optional<String> remoteBranch = remoteBranch();
+      if (!remoteBranch.isPresent()) {
+        return AheadBehind.NO_REMOTE;
+      }
+      fetch(remoteBranch.get());
+      String ahead = runQuietGitCommand(canonical, nativeGitTimeoutInMs, "rev-list --right-only --count " + remoteBranch.get() + "..." + getBranchName());
+      String behind = runQuietGitCommand(canonical, nativeGitTimeoutInMs, "rev-list --left-only --count " + remoteBranch.get() + "..." + getBranchName());
+      return AheadBehind.of(ahead, behind);
+    } catch (Exception e) {
+      throw new GitCommitIdExecutionException("Failed to read ahead behind count: " + e.getMessage(), e);
+    }
+  }
+
+  private Optional<String> remoteBranch() {
+    try {
+      String remoteRef = runQuietGitCommand(canonical, nativeGitTimeoutInMs, "symbolic-ref -q HEAD");
```

Comment:
I had no idea about that property, my bad. Will need to figure out a way to handle it. If I run the command towards the next to last commit on the branch, the command does not return anything. So then it won't be possible to read ahead and behind status.

Reply:
No problem.
I think the fix would be truly trivial with
```
String remoteRef = runQuietGitCommand(canonical, nativeGitTimeoutInMs, "symbolic-ref -q " + evaluateOnCommit);
```

That symbolic-ref call is also being used for getBranch (https://github.com/git-commit-id/maven-git-commit-id-plugin/blob/master/src/main/java/pl/project13/maven/git/NativeGitProvider.java#L100)...that mes me wonder why do we have the `-q` argument here. Maybe there is a reason...not sure.

Reply:
Updated to use evaluateOnCommit. The problem with that is that these errors occur:
> git symbolic-ref -q HEAD^1
fatal: No such ref: HEAD^1
> git symbolic-ref -q HEAD~1
fatal: No such ref: HEAD~1
> git symbolic-ref -q 681210a1ec9d6172227a447af8b4cd8a5ce94fa8
"no output"

So not really sure how to use that parameter.
Documentation:
"In general this property can be set to something generic like `HEAD^1` or point to a branch or tag-name. To support any kind or use-case this configuration can also be set to an entire commit-hash or it's abbreviated version."

Has this changed somehow?

Reply:
I could not use getBranch() either, since that strips refs/heads/ from the branch name and then the next command won't work.

Reply:
Ouch....you seem absolutely right. I'm surprised we never noticed this before. 
Well you now discovered an existing bug :/

The `evaluateOnCommit` thing should really allow everything like `HEAD`, `HEAD^1` or a specific branch name (e.g. `branch`).

Our existing integration-tests do actually not verify that the branch name is correct:
See [here](https://github.com/git-commit-id/maven-git-commit-id-plugin/blob/master/src/test/java/pl/project13/maven/git/GitCommitIdMojoIntegrationTest.java#L1408) and [here](https://github.com/git-commit-id/maven-git-commit-id-plugin/blob/master/src/test/java/pl/project13/maven/git/GitCommitIdMojoIntegrationTest.java#L1441)....

I did some tests locally and it seems that the `symbolic-ref` truly works for `HEADS`, which is not what we want. Giving the command two arguments is also not what we want (e.g. `Given two arguments, creates or updates a symbolic ref <name> to point at the given branch <ref>.`).
I was digging through the internet and came across `git rev-parse --abbrev-ref BRANCH`. That at least seems to work for branches, but does not work for commit hashes and also would indicate tags :-(
Maybe we need `git show-ref --heads ${evaluateOnCommit}`? But that also does not seem to work with commit hashes pointing to branches.
Actually not sure what command we could use, I thought it would be easier to tell a branch name based on an commitish :/

Reply:
Also I quite don't understand why we don't use `--short` for getBranch(), but that would be a different question, that might become obsolete, because the implementation is not doing what we want / need.

Reply:
Maybe `git name-rev --name-only ${evaluateOnCommit}` works quite in the direction. But that would also give a result for `HEAD^1`, but maybe we just could check the output it it contains `~` or `^`...would need to look into that more later.

Reply:
Thinking more about this, it seems reasonable to track this problem in a separate issue.
That would increase the visibility.
I could merge this as is.
WDYT?

Reply:
I agree, should be a separate issue. I can create and start looking into it.
About this issue, if you are ok to merge, then I'm fine with it.